### PR TITLE
End all active spans if the current session span is termianted because of a crash

### PR DESCRIPTION
**Pull request by bidetofevil**
## Goal

End all active spans when a crash happens. Also changed the testing infra to make testing this and other things easier.

## Testing

Added unit tests. Integration tests will come later

### CODE REVIEW

**Comment by bidetofevil** [comment]
* **#615** <a href="https://app.graphite.dev/github/pr/embrace-io/embrace-android-sdk/615?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @bidetofevil and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: embrace-android-sdk/src/test/java/io/embrace/android/embracesdk/internal/spans/SpanRepositoryTest.kt [hunk 1]
```diff
@@ -90,4 +93,19 @@ internal class SpanRepositoryTest {
         assertNull(repository.getSpan(checkNotNull(completedSpan.spanId)))
         assertEquals(0, repository.getCompletedSpans().size)
     }
+
+    @Test
+    fun `active spans become failed and complete when they are forced to fail`() {
+        val startedSpan = FakePersistableEmbraceSpan.started()
+        repository.trackStartedSpan(startedSpan)
+
+        assertSame(startedSpan, repository.getActiveSpans().single())
```
Comment (bidetofevil) [inline embrace-android-sdk/src/test/java/io/embrace/android/embracesdk/internal/spans/SpanRepositoryTest.kt:102]: Not overdoing this as I'm checking out a fake after all.

File: embrace-android-sdk/src/test/java/io/embrace/android/embracesdk/internal/spans/SpanRepositoryTest.kt [hunk 2]
```diff
@@ -90,4 +93,19 @@ internal class SpanRepositoryTest {
...
+        assertFalse(startedSpan.stopped)
```
Comment (bidetofevil) [inline embrace-android-sdk/src/test/java/io/embrace/android/embracesdk/internal/spans/SpanRepositoryTest.kt:103]: Not overdoing this as I'm checking out a fake after all.

File: embrace-android-sdk/src/test/java/io/embrace/android/embracesdk/internal/spans/CurrentSessionSpanImplTests.kt [hunk 3]
```diff
@@ -156,6 +159,49 @@ internal class CurrentSessionSpanImplTests {
         }
     }
 
+    @Test
+    fun `crashing results in the session span and active spans being terminated`() {
```
Comment (fractalwrench) [inline embrace-android-sdk/src/test/java/io/embrace/android/embracesdk/internal/spans/CurrentSessionSpanImplTests.kt:163]: Future work, but is this something that is feasible to cover with an integration test that verifies the payload contains a snapshot of active spans? It should be possible to simulate a crash by getting the uncaught exception handler & manually sending an exception

File: embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/internal/spans/SpanRepository.kt [hunk 4]
```diff
@@ -62,6 +64,18 @@ internal class SpanRepository {
      */
     fun getCompletedSpans(): List<EmbraceSpan> = synchronized(spanIdsInProcess) { completedSpans.values.toList() }
 
+    /**
+     * Stop the existing active spans and mark them as failed
+     */
+    fun failActiveSpans(failureTimeMs: Long) {
+        getActiveSpans().filterNot { it.hasEmbraceAttribute(EmbType.Ux.Session) }.forEach { span ->
```
Comment (fractalwrench) [inline embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/internal/spans/SpanRepository.kt:71]: Does the session span not get marked as stopped during a crash?

**Review by fractalwrench — APPROVED** [review]
LGTM

**Pull request merged**